### ISSUE TITLE
questions regarding optimization and neural network architectures

### ISSUE THREAD

**Issue by martinjankowiak** [YosefLab/scvi-tools#819]
hello,

i have a few somewhat open-ended questions about scVI-style models, especially as regards best practices for model structure and tricks for effective training. (i've tried to figure out what's happening from the codebase but it's not always clear to me what code path is followed where). @romain-lopez suggested this would be a good place to post.

to be concrete let's suppose we're running a scANVI model on two or more datasets, some of which are labeled and some of which are not. 

**[batch/dataset annotations s]:** it's not clear to me if every component distribution of p and q are fed (i.e. conditioned on) s. naively i might expect that p(x|z2, s) should be fed s but that e.g. p(z2|z1) shouldn't be fed s because i want the latent space to be shared (in particular s-independent) across batches. what modeling choice do you make here?

**[batch norm]:** it seems you make extensive use of batch norm. also as far as i can tell you feed your models mini-batches that are pure, i.e. composed of data from a single (as opposed to multiple) dataset. doesn't this have the potential to lead to weird and/or undesired behavior? you might imagine that it'd be necessary to keep track of different batchnorm running buffers for different annotations s. did you ever try anything like that?

**[alternating training procedure]:** it seems you alternate between training of the vanilla ELBO and an auxiliary classifier loss. can you comment on the need for this? in particular it seems you train the classifier using deterministic (i.e. non-stochastic) latent reps as inputs and allow the gradient to also flow through the encoder, is that right?

**[unsupervised pre-training and KL annealing]** do you find that these are important in practice? if so what's your intuition for why?

**[dropout]** it seems you only use dropout in the encoders. is that right? why is this choice made?

**[q(y|z2) instead of q(y|x)]** it seems that in contrast to kingma et al you use a classifier that is conditioned on z2 instead of x. is this simply because x is high-dimensional and so it's probably easier to let q use the denoising that already needs to be learned by q(z2|x)? or is there another reason?

i appreciate that not all of these questions will have clear cut answers but i'd appreciate any intuition!

**Comment by romain-lopez** [comment]
Hi Martin, 

Thanks for those questions. I'll try to answer them here. 

[batch/dataset annotations s]: Yes, I had this question several times in the last years. Basically, we would like to remove s from the latent space. So indeed, we use s only at the level of modeling x. Then, we never add s in the variational network, mainly because we expect it to be a redundant information with x. Indeed, it is obvious to separate batches linearly in gene expression space. You might have better performance doing otherwise, but we did not have a clear benchmark or time to further investigate.

[batch norm]: We had noticed early on that indeed batch norm helps stabilizing the learning of the model and consistently getting reasonable results on a single dataset. We then moved to several batches and kept this. I agree it might have some pathological behaviors, for example in the semi-supervised case. However, in the vanilla scVI minibatches are composed of randomly sampled cells across biological batches so I wouldn't expect this issue. We have coded for gimVI a conditional batchnorm layer (which is what you are suggesting), but we did not notice clear improvements. As for the previous point, we are lacking systematic benchmarking to be able to give a definite answer on this, but you raise a good point.

[alternating training procedure, pretraining and KL]: We went through many different different configurations at early stages of the developments and settled on this one that provided reasonable results. Also, we found the penalized loss introduced by Kingma to be sensitive to the choice of the hyperparameter. However, I wasn't aware that the gradient also flows through the encoder, that might not be entirely desirable actually. About the KL annealing, we found that it sometimes helps getting more cell type separation in latent space, but this is pretty empirical.

[dropout]: I wasn't aware of that either, I am not sure this is desirable as well. Thanks for pointing it out.

[q(y|z2) instead of q(y|x)] : Yes indeed, I think we wanted to classify based on the latent space directly. This seems to be the only reason for me.

**Comment by Munfred** [comment]
Thanks for the responses, I learned something too! If some of those are frequent questions, perhaps consider starting an FAQ and link to it on the landing page at https://www.scvi-tools.org/en/stable/? 

For example, here is my personal FAQ on scVI that I share with my friends (feel free to use it): https://docs.google.com/document/d/12b20khTM2JmB2P_8bMnnbg0RiYE9dxFqMATLglI56EM/edit?usp=sharing

**Comment by romain-lopez** [comment]
This is all great! Actually Adam is setting up a discourse server so we can
all make posts with those type of questions

https://discourse.scvi-tools.org/

Best,
Romain

**Comment by Munfred** [comment]
Awesome! I love discourse!

**Comment by adamgayoso** [comment]
Closing due to inactivity. Happy to continue on our discourse site.
